[#239] Suggest storing the contract metadata on IPFS (#253)
tqtezos/baseDAO: docs/specification.md

This contract implements [TZIP-016](https://gitlab.com/tzip/tzip/-/blob/21fb73fe01df8a744c9b03303e3d73b0f2265eb2/proposals/tzip-16/tzip-16.md).

The DAO contract itself won't store the metadata, rather a file on IPFS (suggested), a dedicated contract or another external storage will contain that.

## Deployment with metadata

> The deployment of contract with metadata has an extra step either:

Reviewer: ```suggestion
The deployment of contract with metadata has an extra step, either:
```

Author: Fixed.

This contract implements [TZIP-016](https://gitlab.com/tzip/tzip/-/blob/21fb73fe01df8a744c9b03303e3d73b0f2265eb2/proposals/tzip-16/tzip-16.md).

> The DAO contract itself won't store the metadata, rather a dedicated contract will contain that.
> The metadata can also be uploaded to IPFS file storage.

Reviewer: ```suggestion
The DAO contract itself won't store the metadata, rather a file on IPFS (suggested), a dedicated contract or another external storage will contain that.
```

Author: Updated.

Motivation:

> * The baseDAO contract is very large and already over the Tezos hard limits.

Reviewer: We forgot to update this, it seems, the contract is no longer over the limit.

Author: I've removed that part.

The deployment of contract with metadata has an extra step: a dedicated contract for carrying metadata has to be originated first.
Then the baseDAO contract should include the reference to a metadata key in the contract in order to be compliant with TZIP-016.

> ## Delployment with metadata that is on IPFS

Reviewer 2: >Delployment

Possible typo.

Author: Updated.

> ## Delployment with metadata that is on IPFS

Reviewer: I think that this could be merged with the previous section, after all it's either way about deploying with metadata and the second step (include the reference) is the same in both cases (except for the different URI).

Author: Make sense, I've merged them now.

Before deploying the contract, the metadata should be uploaded to IPFS. Then the baseDAO contract should include the reference to the

> the metadata CID in the contract in order to be compliant with TZIP-016.

Reviewer 2: > CID

What is this? Wonder if we should include this as an optional argument to storage creation `make` command to optionally include it in storage?

Author: I believe IPFS URI is call CID: https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-16/tzip-16.md#the-tezos-storage-uri-scheme

We can already set this via the `metadata_map = Big_map.literal [("", ipfs://{cid})].

Reviewer 2: Thanks.

>We can already set this via the `metadata_map = Big_map.literal [("", ipfs://{cid})].

Does it make sense to include the handling of this in make file is what I am wondering. Basically, an optional `metadata_cid` argument, if provided, should be included as you shared here. 

Not sure if that is worth the trouble though. @pasqu4le ?

Reviewer: > I believe IPFS URI is call CID:

That doesn't seem to be _quite_ right, AFAICT from the [documentation](https://docs.ipfs.io/how-to/address-ipfs-on-web/#native-urls) the URI is:
> ipfs://{immutable-root}/path/to/resourceA

CID may be a synonym for the `immutable-root` here, but there's also the rest of the path in the URI.

> We can already set this via the `metadata_map = Big_map.literal [("", ipfs://{cid})].

Ah there's a detail here to be careful with, the values for the `metadata_map` need to be `bytes`, so we need an encoding (which is not the same as the `Bytes,pack`/`PACK` conversion, so we cannot do it in LIGO), see [the spec for URI scheme](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-16/tzip-16.md#the-sha256-uri-scheme):
> the encoding of the values must be the direct stream of bytes of the data being stored.
> For instance, an URI starting with `http:` will start with the 5 bytes `0x687474703a` (`h` is `0x68`, `t` is `0x74`, etc.). 

> Basically, an optional metadata_cid argument, if provided, should be included as you shared here.

Well, since we still retain using an external contract as an option what we could do is only to make the encoding above.
So (if there is a common linux command for the conversion) we could get the `metadata_uri` string as an option and encode it in the map.

Reviewer: I think we should do that, given that it fits both the "suggestions" that we give for metadata.

To be clear: I don't think we should modify the behavior of `make metadata` as one might want to add info to the produced JSON (see comments below), but I think the 3 storage generation target should do the encoding instead.

Author: I'm thinking of creating a command (e.g. `make metadata-encoding ipfs_path=..`) when then output `Big_map.literal [("", <some-vytes>]`. Is this what you have in mind?

Reviewer: :thinking: that would mean that the user will need to copy the result of that call to then be used in the storage `make` call.

I was thinking more of replacing the `metadata_map` with a string argument, e.g. for trivialDAO (the default would be an empty string):
```make
$(OUT)/trivialDAO_storage.tz : metadata_uri = ""
```
and convert it for LIGO directly:
```
          ; metadata_map = (Big_map.literal [(\"\", $(call encode_metadata_uri,$(metadata_uri)))] : metadata_map)\
```
where `encode_metadata_uri` is the encoding function that needs to be added.
